### PR TITLE
add annotations

### PR DESCRIPTION
**Pull request by saurabh-jogalekar**

### CODE REVIEW

File: pkg/controller/kubedirectorcluster/cluster.go [hunk 1]
```diff
@@ -158,6 +158,37 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
...
+	} else {
+		cr.Annotations[shared.ConnectionsChanged] = "false"
+
+		shared.LogInfo(
+			reqLogger,
+			cr,
+			shared.EventReasonCluster,
+			"Annotation initialized to false",
+		)
+	}
```
Comment (saurabh-jogalekar) [inline pkg/controller/kubedirectorcluster/cluster.go]: Add the flag in annotations

File: pkg/controller/kubedirectorcluster/cluster.go [hunk 2]
```diff
@@ -211,6 +242,33 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
...
+				shared.LogInfo(
+					reqLogger,
+					cr,
+					shared.EventReasonCluster,
+					"Updated context",
+				)
+			}
+
+		}
+
```
Comment (saurabh-jogalekar) [inline pkg/controller/kubedirectorcluster/cluster.go]: If the hash has changed, set the flag to `true`

File: pkg/controller/kubedirectorcluster/members.go [hunk 3]
```diff
@@ -280,10 +311,59 @@ func handleReadyMembers(
...
+				if cmdErr != nil {
+					shared.LogInfo(
+						reqLogger,
+						cr,
+						shared.EventReasonCluster,
+						fmt.Sprintf("Exited with errors: %s", cmdErr.Error()),
+					)
+				}
+
+			}
```
Comment (saurabh-jogalekar) [inline pkg/controller/kubedirectorcluster/members.go]: call startscript if the flag is `true`

File: pkg/controller/kubedirectorcluster/members.go [hunk 4]
```diff
@@ -280,10 +311,59 @@ func handleReadyMembers(
...
+
+	cr.Annotations[shared.ConnectionsChanged] = "false"
+	if shared.Update(context.TODO(), cr) == nil {
+		shared.LogInfo(
+			reqLogger,
+			cr,
+			shared.EventReasonCluster,
+			"Updated context to false",
+		)
+	}
```
Comment (saurabh-jogalekar) [inline pkg/controller/kubedirectorcluster/members.go]: unset the flag after all roles are done running the startscript

File: pkg/controller/kubedirectorcluster/members.go [hunk 5]
```diff
@@ -49,8 +49,23 @@ func syncMembers(
 	cr *kdv1.KubeDirectorCluster,
 	roles []*roleInfo,
 	configmetaGenerator func(string) string,
+	connChg bool,
```
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/members.go]: nit: let's rename this to hasConnChanged or something more booleny name

File: pkg/controller/kubedirectorcluster/members.go [hunk 6]
```diff
@@ -238,8 +253,43 @@ func handleReadyMembers(
 	cr *kdv1.KubeDirectorCluster,
 	role *roleInfo,
 	configmetaGenerator func(string) string,
+	connChg bool,
```
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/members.go]: I am confused about the use of this boolean, if we have cr then we know member's version as well as the current connection version, why do we need connChg at all ?

File: pkg/shared/types.go [hunk 7]
```diff
@@ -60,3 +60,8 @@ const (
 const (
 	ConnectionsIncrementor = KdDomainBase + "/connUpdateCounter"
 )
+
+// variable for checking if connections change
+// const (
+// 	ConnectionsVersion = KdDomainBase + "/connectionsVersion"
```
Comment (kmathur2) [inline pkg/shared/types.go]: nit:remove this

File: pkg/controller/kubedirectorcluster/types.go [hunk 8]
```diff
@@ -95,6 +95,11 @@ const (
 	2> /opt/guestconfig/configure.stderr  &`
 	fileInjectionCommand = `mkdir -p %s && cd %s &&
 	curl -L %s -o %s`
+	appPrepConfigRunCmdTest = `rm -f /opt/guestconfig/configure.* &&
```
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/types.go]: will we lose previous guest config logs by removing this?

File: pkg/controller/kubedirectorcluster/members.go [hunk 9]
```diff
@@ -238,8 +253,43 @@ func handleReadyMembers(
...
+	// shared.LogInfo(
+	// 	reqLogger,
+	// 	cr,
+	// 	shared.EventReasonCluster,
+	// 	fmt.Sprintf("In sync members, Annotations chg: %s", cr.Annotations[shared.ConnectionsChanged]),
+	// )
+	var connectionsVersion int64
+
+	if connectionsVersionStr, ok := cr.Annotations[shared.ConnectionsIncrementor]; ok {
+		connectionsVersion, _ = strconv.ParseInt(connectionsVersionStr, 10, 64)
```
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/members.go]: wondering what happens if someone tampers with annotation and either deletes it Or put a nonint value . We shoud test that scenario

File: pkg/apis/kubedirector/v1beta1/kubedirectorcluster_types.go [hunk 10]
```diff
@@ -183,6 +183,7 @@ type MemberStateDetail struct {
 	LastConfiguredContainer  string              `json:"lastConfiguredContainer,omitempty"`
 	LastKnownContainerState  string              `json:"lastKnownContainerState,omitempty"`
 	PendingNotifyCmds        []*NotificationDesc `json:"pendingNotifyCmds,omitempty"`
+	LastConnectionVersion    int64               `json:"lastConnectionVersion,omitempty"`
```
Comment (kmathur2) [inline pkg/apis/kubedirector/v1beta1/kubedirectorcluster_types.go]: *int64 to accommodate nil

File: deploy/example_clusters/cr-cluster-jupyter-notebook.yaml [hunk 11]
```diff
@@ -4,12 +4,15 @@ metadata:
   name: "jupyter-notebook"
 spec:
   app: jupyter-notebook
```
Comment (kmathur2) [inline deploy/example_clusters/cr-cluster-jupyter-notebook.yaml]: remove test-secret and cpu val changes

File: deploy/example_catalog/cr-app-training-engine.json [hunk 12]
```diff
@@ -54,23 +54,23 @@
         "roles": [{
                 "cardinality": "1",
                 "configPackage": {
```
Comment (kmathur2) [inline deploy/example_catalog/cr-app-training-engine.json]: revert these changes

File: deploy/example_catalog/cr-app-jupyter-notebook.json [hunk 13]
```diff
@@ -55,7 +55,7 @@
             }
         ],
         "defaultConfigPackage": {
-            "packageURL": "file:///opt/configscript/appconfig.tgz"
+            "packageURL": "http://16.143.21.166/thirdparty/k8sapps/hooks/bluedata-catalog/appconfig.tgz"
```
Comment (kmathur2) [inline deploy/example_catalog/cr-app-jupyter-notebook.json]: revert

File: pkg/controller/kubedirectorcluster/members.go [hunk 14]
```diff
@@ -280,10 +330,83 @@ func handleReadyMembers(
...
+				cmd := fmt.Sprintf(appPrepConfigRunCmdTest, containerID)
+
+				shared.LogInfo(
+					reqLogger,
+					cr,
+					shared.EventReasonCluster,
+					fmt.Sprintf("Executing command %s", cmd),
+				)
+				cmdErr := executor.RunScript(
+					reqLogger,
```
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/members.go:348]: is the configmeta successfully updated by now?  how do we know?

File: pkg/controller/kubedirectorcluster/members.go [hunk 15]
```diff
@@ -280,10 +330,83 @@ func handleReadyMembers(
...
+					m.StateDetail.LastConfiguredContainer,
+					executor.AppContainerName,
+					"app config",
+					strings.NewReader(cmd),
+				)
+				if cmdErr != nil {
+					shared.LogInfo(
+						reqLogger,
+						cr,
+						shared.EventReasonCluster,
```
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/members.go]: if we error out, should we put member in error state? we do in expansion case I think but should we put here? what do you think?

File: pkg/controller/kubedirectorcluster/members.go (hunk 9 above)
Comment (kmathur2) [inline pkg/controller/kubedirectorcluster/members.go]: if this errors out we should atleast log, may be put it back to member's number

File: secret.yaml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: Secret
+metadata:
```
Comment (kmathur2) [inline secret.yaml]: remove

File: pkg/shared/types.go [hunk 17]
```diff
@@ -60,3 +60,8 @@ const (
 const (
 	ConnectionsIncrementor = KdDomainBase + "/connUpdateCounter"
```
Comment (kmathur2) [inline pkg/shared/types.go]: we can rename this to reflect it as a connection version

File: pkg/controller/kubedirectorcluster/members.go (hunk 6 above)
Comment (saurabh-jogalekar) [reply pkg/controller/kubedirectorcluster/members.go]: we dont need this flag. will delete the code

File: pkg/controller/kubedirectorcluster/types.go (hunk 8 above)
Comment (saurabh-jogalekar) [reply pkg/controller/kubedirectorcluster/types.go]: currently, yes

File: pkg/controller/kubedirectorcluster/members.go (hunk 14 above)
Comment (saurabh-jogalekar) [reply pkg/controller/kubedirectorcluster/members.go:348]: we know that because if there is an error in configmeta, we [return](https://github.com/saurabh-jogalekar/kubedirector/blob/10bfba8e9d1a5f76a72b29a6b6f8dd33a0da771f/pkg/controller/kubedirectorcluster/members.go#L331)

File: pkg/controller/kubedirectorcluster/members.go (hunk 15 above)
Comment (saurabh-jogalekar) [reply pkg/controller/kubedirectorcluster/members.go]: the member now goes in the same state as it would be when configmeta doesn't get injected, albeit with a different error message
